dot/state/grandpa_changes.go

```diff
@@ -0,0 +1,313 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package state
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+type conditionFunc[T any] func(T) (bool, error)
+type isDescendantOfFunc func(parent, child common.Hash) (bool, error)
+
+type pendingChange struct {
+	bestFinalizedNumber uint32
+	delay               uint32
+	nextAuthorities     []types.Authority
+	announcingHeader    *types.Header
+}
+
+func (p pendingChange) String() string {
+	return fmt.Sprintf("announcing header: %s (%d), delay: %d, next authorities: %d",
+		p.announcingHeader.Hash(), p.announcingHeader.Number, p.delay, len(p.nextAuthorities))
+}
+
+func (p *pendingChange) effectiveNumber() uint {
+	return p.announcingHeader.Number + uint(p.delay)
+}
+
+type orderedPendingChanges []*pendingChange
```

Comment:
why `[]*pendingChange` vs `[]pendingChange`.  `pendingChange` doesn't contain a lot of data, so I don't see the harm in just copying the type and reducing complexity when dealing with a slice of pendingChanges.